beacon/light/committee_chain.go

```diff
@@ -0,0 +1,587 @@
+	reorg := oldRoot != (common.Hash{}) && oldRoot != update.NextSyncCommitteeRoot
+	if oldUpdate, ok := s.updates.get(period); ok && !update.Score().BetterThan(oldUpdate.Score()) {
+		// a better or equal update already exists; no changes, only fail if new one tried to reorg
+		if reorg {
+			return ErrCannotReorg
+		}
+		return nil
+	}
+	if s.fixedRoots.periods.Includes(period+1) && reorg {
+		return ErrCannotReorg
+	}
+	if !s.verifyUpdate(update) {
+		return ErrInvalidUpdate
+	}
+	addCommittee := !s.committees.periods.Includes(period+1) || reorg
+	if addCommittee {
+		if nextCommittee == nil {
+			return ErrNeedCommittee
+		}
+		if nextCommittee.Root() != update.NextSyncCommitteeRoot {
+			return ErrWrongCommitteeRoot
+		}
+	}
+	batch := s.db.NewBatch()
+	if reorg {
+		s.rollback(batch, period+1)
+	}
+	if addCommittee {
+		if err := s.committees.add(batch, period+1, nextCommittee); err != nil {
+			return err
+		}
+		s.syncCommitteeCache.Remove(period + 1)
+	}
+	if err := s.updates.add(batch, period, update); err != nil {
+		return err
+	}
+	if err := batch.Write(); err != nil {
+		log.Error("Error writing batch into chain database", "error", err)
+		return err
+	}
+	log.Info("Inserted new committee update", "period", period, "next committee root", update.NextSyncCommitteeRoot)
+	return nil
+}
+
+// NextSyncPeriod returns the next period where an update can be added and also
+// whether the chain is initialized at all.
+func (s *CommitteeChain) NextSyncPeriod() (uint64, bool) {
+	s.chainmu.RLock()
+	defer s.chainmu.RUnlock()
+
+	if s.committees.periods.IsEmpty() {
+		return 0, false
+	}
+	if !s.updates.periods.IsEmpty() {
+		return s.updates.periods.AfterLast, true
+	}
+	return s.committees.periods.AfterLast - 1, true
+}
+
+// rollback removes all committees and fixed roots from the given period and updates
+// starting from the previous period.
+func (s *CommitteeChain) rollback(batch ethdb.Batch, period uint64) {
+	s.deleteCommitteesFrom(batch, period)
+	s.fixedRoots.deleteFrom(batch, period)
+	if period > 0 {
+		period--
+	}
+	s.updates.deleteFrom(batch, period)
+}
+
+// getCommitteeRoot returns the committee root at the given period, either fixed,
+// proven by a previous update or both. It returns an empty hash if the committee
+// root is unknown.
+func (s *CommitteeChain) getCommitteeRoot(period uint64) common.Hash {
+	if root, ok := s.fixedRoots.get(period); ok || period == 0 {
+		return root
+	}
+	if update, ok := s.updates.get(period - 1); ok {
+		return update.NextSyncCommitteeRoot
+	}
+	return common.Hash{}
+}
+
+// getSyncCommittee returns the deserialized sync committee at the given period.
+func (s *CommitteeChain) getSyncCommittee(period uint64) syncCommittee {
+	if c, ok := s.syncCommitteeCache.Get(period); ok {
+		return c
+	}
+	if sc, ok := s.committees.get(period); ok {
+		c, err := s.sigVerifier.deserializeSyncCommittee(sc)
+		if err != nil {
+			log.Error("Sync committee deserialization error", "error", err)
```

Comment:
I think we should bubble up the error, propagate it to the topmost call and somehow print the error instead of erroring here.

Reply:
done